library/src/main/java/com/google/android/exoplayer/extractor/ts/MpaReader.java

```diff
@@ -0,0 +1,293 @@
+     *
+     * @param pesBuffer The buffer whose position should be advanced.
+     * @return True if a sync word position was found. False otherwise.
+     */
+    private boolean skipToNextSync(ParsableByteArray pesBuffer) {
+        byte[] mpaData = pesBuffer.data;
+        int startOffset = pesBuffer.getPosition();
+        int endOffset = pesBuffer.limit();
+        for (int i = startOffset; i < endOffset - 1; i++) {
+            int syncBits = ((mpaData[i] & 0xFF) << 8 ) | (mpaData[i + 1] & 0xFF);
+            if ((syncBits & 0xFFF0) == 0xFFF0) {
+                hasCrc = (mpaData[i + 1] & 0x1) == 0;
+                pesBuffer.setPosition(i);
+                return true;
+            }
+        }
+        pesBuffer.setPosition(endOffset);
+        return false;
+    }
+
+    /**
+     * Calculates MPEG Audio frame size
+     *
+     * @param layer The MPEG layer
+     * @param LSF Low Sample rate Format (MPEG 2)
+     * @param bitrate The bitrate in bits per second
+     * @param samplesPerSec The sampling rate in hertz
+     * @param -paddingSize
+     * @return Frame size in bytes
+     */
+    private static int CalcMpaFrameSize (int layer, int LSF, int bitrate, int samplesPerSec, int paddingSize) {
+        return (int)(Math.floor(MPA_COEFFICIENTS[LSF][layer] * bitrate / samplesPerSec) + paddingSize) * MPA_SLOT_SIZE[layer];
+    }
+
+    /**
+     * Parses the sample header.
+     */
+    private void parseHeader() {
+        int headerLength = hasCrc ? HEADER_SIZE + CRC_SIZE : HEADER_SIZE;
+
+        if (!hasOutputFormat) {
+            mpaScratch.setPosition(0);
+            mpaScratch.skipBits(12);
+            int isLSF = (!mpaScratch.readBit()) ? 1 : 0;
+            int layer = mpaScratch.readBits(2) ^ 3;
+            mpaScratch.skipBits(1);
+            int audioObjectType = 32 + layer;
+            int bitRate = MPA_BITRATES[isLSF][layer][mpaScratch.readBits(4)];
+            int sampleRate = MPA_SAMPLING_RATES[3 - isLSF][mpaScratch.readBits(2)];
+            int sampleRateIndex = CodecSpecificDataUtil.getSampleRateIndex(sampleRate);
+            int paddingBit = (mpaScratch.readBit()) ? 1 : 0;
+            mpaScratch.skipBits(1);
+            int channelConfig = mpaScratch.readBits(2) == 3 ? 1 : 2;
+
+            byte[] audioSpecificConfig = CodecSpecificDataUtil.buildAudioSpecificConfig(
+                    audioObjectType, sampleRateIndex, channelConfig);
+            Pair<Integer, Integer> audioParams = CodecSpecificDataUtil.parseAudioSpecificConfig(
+                    audioSpecificConfig);
+
+            // need to investigate how to detect if the mpeg decoder supports Layers other than Layer III
+            MediaFormat mediaFormat = MediaFormat.createAudioFormat(/*isLSF == 1 ?*/ MimeTypes.AUDIO_MPEG/* : MimeTypes.AUDIO_MP1L2*/,
+                    MediaFormat.NO_VALUE, audioParams.second, audioParams.first,
+                    Collections.singletonList(audioSpecificConfig));
```

Comment:
It must be a codec implementation issue, it works properly in all my boxes
(5 different boxes from 4 different vendors, 2 Android TVs and another 2
TVs with  Android 4)

Sergio
El 16/6/2015 9:24 a. m., "optman" <notifications@github.com> escribió:

> In
> library/src/main/java/com/google/android/exoplayer/extractor/ts/MpaReader.java
> <https://github.com/google/ExoPlayer/pull/325#discussion_r32496431>:
>
> > +            int bitRate = MPA_BITRATES[isLSF][layer][mpaScratch.readBits(4)];
> > +            int sampleRate = MPA_SAMPLING_RATES[3 - isLSF][mpaScratch.readBits(2)];
> > +            int sampleRateIndex = CodecSpecificDataUtil.getSampleRateIndex(sampleRate);
> > +            int paddingBit = (mpaScratch.readBit()) ? 1 : 0;
> > +            mpaScratch.skipBits(1);
> > +            int channelConfig = mpaScratch.readBits(2) == 3 ? 1 : 2;
> > +
> > +            byte[] audioSpecificConfig = CodecSpecificDataUtil.buildAudioSpecificConfig(
> > +                    audioObjectType, sampleRateIndex, channelConfig);
> > +            Pair<Integer, Integer> audioParams = CodecSpecificDataUtil.parseAudioSpecificConfig(
> > +                    audioSpecificConfig);
> > +
> > +            // need to investigate how to detect if the mpeg decoder supports Layers other than Layer III
> > +            MediaFormat mediaFormat = MediaFormat.createAudioFormat(/*isLSF == 1 ?*/ MimeTypes.AUDIO_MPEG/* : MimeTypes.AUDIO_MP1L2*/,
> > +                    MediaFormat.NO_VALUE, audioParams.second, audioParams.first,
> > +                    Collections.singletonList(audioSpecificConfig));
>
> I have try the links you provided, it crash when initialData is not set to
> null.
>
> http://www.endlesspools.com/iphone/iphone_streams/Fastlane/Fastlane_128k.m3u8
> http://www2.lapresse.it/video/mp4/1.15753/1.15753_128k.m3u8
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/ExoPlayer/pull/325/files#r32496431>.
>